### PR TITLE
Add point-geometry and vector-tile types

### LINKED ISSUE #788: Remove vector-tile module declaration
## Motivation

After fixing #787 I wanted to remove the not so optional solution with the vector tile typings I did.
@webberig will hopefully make a PR to definitely typed repo, after that the module declaration of mapbox vector tile can be removed from the code and the post build step.

This issue is to follow on the progress of this.

### PR DESCRIPTION
**Pull request by HarelM**
## Launch Checklist
Fixes #788
Added `point-geometry` and `vector-tile` types packages.

Relevant things to consider:
1. camera.ts#580 - this is not the same, it used to allow sending `{x:number, y:number}` object, which will break after this, I'm not sure I mind as the docs doesn't really support this. I removed the test around it. I can bring this code back, but it will need an API change of the object that is being sent, which again, feels wrong.
2. Removed `Point.ts` and reverted to use `point-geomerty` since `vector-tile` is using it and I wanted to keep the implementation aligned.
3. There's no `register('Point', Point)`, I'm not sure this is a problem, need to check the 1.x branch, but I don't think it was ever there...

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Include before/after visuals or gifs if this PR includes visual changes.

### CODE REVIEW

**Comment by HarelM** [comment]
@wipfli @birkskyum @astridx Ready for review.
All the packages that this project uses are now fully typed! :-)

**Comment by HarelM** [comment]
I think it might fix #790 too, not sure though...

File: src/ui/camera.test.ts [hunk 1]
```diff
@@ -1777,14 +1777,6 @@ describe('#cameraForBounds', () => {
         expect(fixedLngLat(transform.center, 4)).toEqual({lng: -100.5, lat: 44.4717});
     });
 
-    test('offset as object', () => {
```
Comment (birkskyum) [inline src/ui/camera.test.ts:1780]: All looks really good :) I wonder, how come this test is removed?

**Review by birkskyum — COMMENTED** [review]
Looks really good :)

File: src/ui/camera.test.ts (hunk 1 above)
Comment (HarelM) [reply src/ui/camera.test.ts:1780]: It doesn't fit the API. offset should be PointLike, which is not an x,y object.
I wonder why typescript didn't complain here...
I might be willing though.
The test failed, as the code was changed. You can see how this line looked in 1.x branch.

**Review by birkskyum — APPROVED** [review]

**Comment by birkskyum** [comment]
I got this message in the tsconfig.json

screen

**Comment by HarelM** [comment]
I removed vector-tile from there, or might be that there's a need to add @mapbix/vector-tile, I'm not sure.
Where is this error coming from? The IDE?

**Comment by birkskyum** [comment]
Yes, the IDE. I don't see any compile errors

File: src/ui/camera.ts [hunk 2]
```diff
@@ -14,6 +14,8 @@ import type {LngLatBoundsLike} from '../geo/lng_lat_bounds';
 import type {TaskID} from '../util/task_queue';
 import type {PaddingOptions} from '../geo/edge_insets';
 
+export type PointLike = Point | [number, number];
```
Comment (wipfli) [inline src/ui/camera.ts:17]: How did you choose `src/ui/camera.ts` to define the type `PointLike`?

File: build/post-ts-build.js [hunk 3]
```diff
@@ -37,15 +37,6 @@ child_process.execSync(`dts-bundle-generator --no-check --umd-module-name=maplib
...
 types = types.replace(/declare class/g, "export declare class");
-// Missing vector-tile types that are added to this file too
-let file = fs.readFileSync("./src/types/packages-types/vector-tile/index.d.ts", 'utf8');
-let lines = file.split("\n").filter(Boolean);
-lines.pop(); // last } is removed too.
-types = lines.join("\n") + "\n" + types;
-// remove imports of vector-tile and declare module lines
-types = types.split("\n")
-    .filter(l => !l.includes("@mapbox/vector-tile"))
-    .join("\n");
```
Comment (wipfli) [inline build/post-ts-build.js:48]: Feels good to remove this :)

File: src/ui/camera.test.ts (hunk 1 above)
Comment (wipfli) [reply src/ui/camera.test.ts:1780]: https://github.com/maplibre/maplibre-gl-js/blob/cb3a4200bfbc113969e9dcbedae2aa7655f9aa98/test/unit/ui/camera.test.js#L1878-L1885

is how this test looked like in the 1.x branch.

Comment (wipfli) [reply src/ui/camera.test.ts:1780]: I think you should add this test back in.

Comment (wipfli) [reply src/ui/camera.test.ts:1780]: The reason why this test fails is that the convert function of `@mapbox/point-geometry`:

https://github.com/mapbox/point-geometry/blob/bf471dc7ceb56299d4f5152c4badbeba193b0903/index.js#L304-L312

```js
Point.convert = function (a) {
    if (a instanceof Point) {
        return a;
    }
    if (Array.isArray(a)) {
        return new Point(a[0], a[1]);
    }
    return a;
};
``` 

does not do the same as our `convert` function that we made in the TypeScript migration I think:

```js
    static convert(a: PointLike | {x: number; y: number}): Point {
        if (a instanceof Point) {
            return a;
        }
        if (Array.isArray(a)) {
            return new Point(a[0], a[1]);
        }
        if (typeof a.x === 'number') {
            return new Point(a.x, a.y);
        }
        throw new Error(`Unable to convert to point: ${JSON.stringify(a)}`);
    }
```

**Review by wipfli — COMMENTED** [review]
Thanks for working on this @HarelM. Other than the convert-discrepancy this looks good to me.

File: src/ui/camera.ts (hunk 2 above)
Comment (HarelM) [reply src/ui/camera.ts:17]: I don't know, felt like the right place. I don't think it matters much...

File: build/post-ts-build.js (hunk 3 above)
Comment (HarelM) [reply build/post-ts-build.js:48]: Yup :-)

**Comment by HarelM** [comment]
I moved one line from camera.js to point convert as it felt like it should be there when migrating to typescript.
But from API point of view in terms of typescript we don't really support offset as x,y object although the code used to support it.
IDK...

File: src/ui/camera.test.ts (hunk 1 above)
Comment (wipfli) [reply src/ui/camera.test.ts:1780]: I opened issue #808. I think that the convert function change should not be a problem and that we should move on and fix the typing issues by merging this pull request.

**Review by wipfli — APPROVED** [review]

**Pull request merged**